kernel/aster-nix/src/syscall/utimens.rs

```diff
@@ -1,91 +1,255 @@
+        timespecs_addr += core::mem::size_of::<timespec_t>();
+        let mut mutime = read_val_from_user::<timespec_t>(timespecs_addr)?;
+
+        autime.nsec = 0;
+        mutime.nsec = 0;
+        Some(Times {
+            atime: autime,
+            mtime: mutime,
+        })
+    } else {
+        None
+    };
+    do_utimes(AT_FDCWD, pathname_ptr, times, 0)
+}
 
-    if atime.is_none() && mtime.is_none() {
-        return Ok(SyscallReturn::Return(0));
+// Structure to hold access and modification times
+#[derive(Debug)]
+struct Times {
+    atime: timespec_t,
+    mtime: timespec_t,
+}
+
+fn vfs_utimes(dentry: &Arc<Dentry>, times: Option<Times>) -> Result<SyscallReturn> {
+    let (atime, mtime) = match times {
+        Some(times) => {
+            if !times.atime.is_valid() || !times.mtime.is_valid() {
+                return_errno_with_message!(Errno::EINVAL, "invalid time")
+            }
+            if times.atime.is_utime_now() && times.mtime.is_utime_now() {
+                let now = RealTimeCoarseClock::get().read_time();
+                (now, now)
+            } else {
+                let atime = if times.atime.is_utime_omit() {
+                    dentry.metadata().atime
+                } else {
+                    Duration::from(times.atime)
+                };
+                let mtime = if times.mtime.is_utime_omit() {
+                    dentry.metadata().mtime
+                } else {
+                    Duration::from(times.mtime)
+                };
+                (atime, mtime)
+            }
+        }
+        None => {
+            let now = RealTimeCoarseClock::get().read_time();
+            (now, now)
+        }
+    };
+
+    // Update times
+    let now = RealTimeCoarseClock::get().read_time();
+    println!("atime: {:?}, mtime: {:?}, now: {:?}", atime, mtime, now);
+    dentry.set_atime(atime);
+    dentry.set_mtime(mtime);
+    dentry.set_ctime(now);
+
+    Ok(SyscallReturn::Return(0))
+}
+
+// Function to update the access and modification times of a file by its file descriptor
+fn do_utimes_fd(fd: FileDesc, times: Option<Times>, flags: UtimensFlags) -> Result<SyscallReturn> {
+    // Check if flags are empty, otherwise return error
+    if !flags.is_empty() {
+        return_errno_with_message!(Errno::EINVAL, "invalid flags")
     }
     let current = current!();
     let dentry = {
-        let path = path.to_string_lossy();
-        if path.is_empty() {
-            return_errno_with_message!(Errno::ENOENT, "path is empty");
+        let file_table = current.file_table().lock();
+        let file = file_table.get_file(fd)?;
+        let inode_handle = file
+            .downcast_ref::<InodeHandle>()
+            .ok_or(Error::with_message(Errno::EBADF, "not inode"))?;
+        inode_handle.dentry().clone()
+    };
+
+    vfs_utimes(&dentry, times)
+}
+
+// Function to update the access and modification times of a file by its path
```

Comment:
You can use https://github.com/asterinas/asterinas/blob/74e4623e70121943eb3112af41f0810dc77d7989/kernel/aster-nix/src/fs/fs_resolver.rs#L326-L328 to simplify the code for "dirfd & pathname".